src/libraries/System.Private.CoreLib/src/System/Collections/HashHelpers.cs

```diff
@@ -90,21 +90,23 @@ public static int ExpandPrime(int oldSize)
         }
 
 #if BIT64
+        // Returns 64-bit multiplicative inverse of the divisor: ceil(2**64 / divisor)
         public static ulong GetFastModMultiplier(uint divisor)
             => ulong.MaxValue / divisor + 1;
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static uint FastMod(uint value, uint divisor, ulong multiplier)
         {
-            // Using fastmod from Daniel Lemire https://lemire.me/blog/2019/02/08/faster-remainders-when-the-divisor-is-a-constant-beating-compilers-and-libdivide/
+            // Using improved Daniel Lemire's fastmod: https://github.com/dotnet/runtime/pull/406
+
+            // Required for the rounding up below to work correctly
+            Debug.Assert(divisor <= int.MaxValue);
 
             ulong lowbits = multiplier * value;
-            // 64bit * 64bit => 128bit isn't currently supported by Math https://github.com/dotnet/corefx/issues/41822
-            // otherwise we'd want this to be (uint)Math.MultiplyHigh(lowbits, divisor)
```

Comment:
When available still want to replace it with `(uint)Math.BigMul(lowbits, divisor, out _)`; since that will be just one multiply operation?

Reply:
@benaadams I will likely put that comment back.  On x64 that may save one `INC` instruction; however, it is not clear whether we want to use `Math.BigMul` on ARM64.  On ARM processors `UMULH` has slightly worse execution latency and throughput than the ordinary multiplication (see the reference to ARM optimization guide in the top post).  The LKK article mentions that in the middle of section 5.1 — they say that might be the reason their method could not beat the compiler on ARM.

Furthermore, under the optimal codegen the extra addition should be fused with the multiplication, therefore, it would be free: `MUL` is an alias for the `MADD` (multiply-add) instruction.  For reference, here is the ARM64 gcc 8.2 output for my `fastmod` C++ function (via godbolt.org):
``` c++
uint32_t fastmod(uint32_t n, uint64_t c, uint32_t d)
{
    uint64_t lowbits = c * n;
    return (((lowbits >> 32) + 1) * d) >> 32;
}
```
```
fastmod(unsigned int n, unsigned long c, unsigned int d):
        uxtw    x0, w0          ; Zero-extend n to 64 bit
        uxtw    x2, w2          ; Zero-extend d to 64 bit
        mul     x0, x0, x1      ; x0 = c * n
        lsr     x0, x0, 32      ; x0 = x0 >> 32
        madd    x0, x0, x2, x2  ; x0 = x0 * d + d (fused multiply-add)
        lsr     x0, x0, 32      ; x0 = x0 >> 32
        ret
```

Reply:
You are right; it should benefit both x64 and ARM64.  I restored the comment.